flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/FileLifeCycleListener.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.api.functions.sink.filesystem;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.core.fs.Path;
+
+/**
+ * Listener about the status of file.
+ */
+@Internal
+public interface FileLifeCycleListener<IN, BucketID> {
+
+	/**
+	 * Notifies a new file has been opened.
+	 *
+	 * <p>Note that this does not mean that the file has been created in the file system. It is
+	 * only created logically and the actual file will be generated after it is committed.
+	 *
+	 * @param bucket The bucket of newly opened file.
+	 * @param newPath The path of newly opened file.
+	 */
+	void openPartFile(Bucket<IN, BucketID> bucket, Path newPath);
```

Comment:
Here we should not expose the whole `Bucket` as this is an internal class that can change anytime. 

We can pass here a new interface, something like `BucketInfo`, which is going to be implemented by the `Bucket` and which only exposes the necessary information. Currently, these are the public methods in the `Bucket`, ie.  the `BucketID getBucketId()`, `Path getBucketPath()`, and `long getPartCounter()`.

Actually, I would recommend to not expose the part counter because this will not be supported anymore in the new `FileSink`. If you need access to the in-progress file path, then simply pass that (the result of the `assembleNewPartPath()`) it in a field and have a method `getInProgressFilePath()` in the new `BucketInfo` interface.

Reply:
Good point!
+1 for not exposing the part counter.
In addition, consider whether it is enough to expose only the `BucketID`? Because the new part path is already provided, I think `BucketPath` is useless in this case too.